codegenerator/cli/templates/static/codegen/src/eventFetching/ChainManager.res

```diff
@@ -188,41 +187,29 @@ let setChainFetcher = (self: t, chainFetcher: ChainFetcher.t) => {
   }
 }
 
-type earliestQueueItem =
-  | ArbitraryEventQueue(Types.eventBatchQueueItem, array<Types.eventBatchQueueItem>)
-  | EventFetchers(Types.eventBatchQueueItem, ChainMap.t<PartitionedFetchState.t>)
-
-let rec getFirstArbitraryEventsItemForChain = (
-  queue: array<Types.eventBatchQueueItem>,
-  ~index=0,
-  ~chain,
-) => {
-  switch queue[index] {
-  | None => None
-  | Some(first) =>
-    let nextIndex = index + 1
-    if first.chain == chain {
-      Some((first, () => queue->Utils.Array.removeAtIndex(index)))
-    } else {
-      queue->getFirstArbitraryEventsItemForChain(~chain, ~index=nextIndex)
-    }
-  }
-}
+let getFirstArbitraryEventsItemForChain = (queue: array<Types.eventBatchQueueItem>, ~chain) =>
+  queue
+  ->Utils.Array.findReverseWithIndex((item: Types.eventBatchQueueItem) => {
+    item.chain == chain
+  })
+  ->Option.map(((item, index)) => {
+    FetchState.item,
+    popItemOffQueue: () => queue->Utils.Array.spliceInPlace(~pos=index, ~remove=1)->ignore,
+  })
```

Comment:
It also creates an object here, so theoretically, the solution with ArraySlice should be even faster. The only challenge would be to pop items somewhere in the middle, as you do here. It'd require additional changes in the queue datastructure, by having a separate queue by chain, so I agree that your current solution is better 👍

Reply:
Yeah, I thought about this. But this case is only hit during unorderd multichain mode after a dynamic contract registration. So we definitely shouldn't be optimizing for this case off the bat.